labgrid/util/agents/network_interface.py

```diff
@@ -0,0 +1,359 @@
+
+
+def connection_from_dict(data):
+    con = NM.SimpleConnection.new()
+    for setting_name, setting_data in data.items():
+        typ = NM.Setting.lookup_type(setting_name).pytype
+        setting = typ()
+        for k, v in setting_data.items():
+            if not v:
+                continue
+
+            if k == 'mac-address':
+                v = ':'.join("%02X" % x for x in v)
+            elif k == 'ssid':
+                if isinstance(v, str):
+                    v = v.encode()
+                v = GLib.Bytes(v)
+            elif k == 'addresses':
+                # setting addresses via the property doesn't seem to work
+                for x in v:
+                    setting.add_address(address_from_str(x))
+                v = None
+
+            if not v:
+                continue
+            try:
+                setting.set_property(k, v)
+            except TypeError as e:
+                raise TypeError("Failed to set propery {}".format(k)) from e
+        con.add_setting(setting)
+    return con
+
+class NMDev:
+    def __init__(self, interface):
+        self._interface = interface
+        self._nm_dev = nm.get_device_by_iface(interface)
+
+    def _delete_connection(self, con):
+        future = Future()
+        def cb(con, res, error):
+            assert error is None
+            try:
+                res = con.delete_finish(res)
+                future.set(res)
+            except Exception as e:
+                future.set(e)
+
+        con.delete_async(
+            None,  # cancellable
+            cb,
+            None,
+        )
+
+    def get_settings(self):
+        lg_con = nm.get_connection_by_id("labgrid-{}".format(self._interface))
+        if lg_con:
+            return dict(lg_con.to_dbus(NM.ConnectionSerializationFlags.ALL))
+
+    def get_active_settings(self):
+        active_con = self._nm_dev.get_active_connection()
+        if active_con:
+            con = active_con.get_connection()
+            return dict(con.to_dbus(NM.ConnectionSerializationFlags.ALL))
+
+    def configure(self, data):
+        lg_con = nm.get_connection_by_id("labgrid-{}".format(self._interface))
+        if lg_con:
+            self._delete_connection(lg_con)
+        data['connection'].update({
+            'id': "labgrid-{}".format(self._interface),
+            'autoconnect': False,
+            'interface-name': self._interface,
+        })
+        con = connection_from_dict(data)
+
+        future = Future()
+        def cb(dev, res, error):
+            assert error is None
+            try:
+                res = nm.add_and_activate_connection_finish(res)
+                future.set(res)
+            except Exception as e:
+                future.set(e)
+
+        nm.add_and_activate_connection_async(
+            con,
+            self._nm_dev,
+            None,  # specific_object
+            None,  # cancellable
+            cb,
+            None,
+        )
+
+        active_con = future.wait()
```

Comment:
Will `active_con` be used in the future?

Reply:
No, I'll add a comment.